## DATA 240: Assignment No: 3

## Binary classification with logistic regression 

Please download heart_disease_train.csv and heart_disease_test.csv. The dataset is for cardiovascular study. The
target variable is ‘TenYearCHD’, which shows whether the patient has 10-year risk of future coronary heart disease
(CHD). You can find a description of the variables in cardiovascular.txt.

• Gender: male(1) or female(0)\
• Age: Age of the patient\
• Current Smoker: whether or not the patient is a current smoker \
• Cigs Per Day: the number of cigarettes that the person smoked on average in one day\
• BP Meds: whether or not the patient was on blood pressure medication \
• Prevalent Stroke: whether or not the patient had previously had a stroke \
• Prevalent Hyp: whether or not the patient was hypertensive \
• Diabetes: whether or not the patient had diabetes \
• Tot Chol: total cholesterol level \
• Sys BP: systolic blood pressure\
• Dia BP: diastolic blood pressure \
• BMI: Body Mass Index \
• Heart Rate: heart rate \
• Glucose: glucose level \
• TenYearCHD: 10 year risk of coronary heart disease (1 means Yes, 0 means No)

In [51]:
heart_train = pd.read_csv('heart_disease_train.csv')

In [52]:
heart_test = pd.read_csv('heart_disease_test.csv')

In [53]:
heart_train.head()

gender  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0       1   63        1.0              0         0.0     0.0                0   
1       1   45        3.0              0         0.0     0.0                0   
2       0   51        1.0              1         2.0     0.0                0   
3       1   45        3.0              1        30.0     0.0                0   
4       0   45        2.0              1         3.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             1         0    190.0  148.0   90.0  27.13       72.0     86.0   
1             0         0    162.0  125.0   89.0  27.98       85.0     83.0   
2             0         0    261.0  127.0   81.0  20.24       75.0     96.0   
3             0         0    250.0  126.0   89.5  28.68       75.0     92.0   
4             0         0    250.0  130.0   80.0  20.24       90.0     86.0   

   TenYearCHD  
0           0  
1           0  
2           0  
3           0  
4           0

In [54]:
heart_test.head()

gender  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0       0   56        1.0              0         0.0     0.0                0   
1       0   46        2.0              0         0.0     0.0                0   
2       0   52        1.0              0         0.0     0.0                0   
3       0   46        3.0              1        23.0     0.0                0   
4       1   69        1.0              1         1.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             1         1    273.0  136.0   80.0  27.73       90.0    210.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    275.0  112.0   71.0  25.68       80.0      NaN   
3             0         0    285.0  130.0   84.0  23.10       85.0     85.0   
4             0         0    245.0  123.0   77.0  26.58       70.0     81.0   

   TenYearCHD  
0           1  
1           0  
2           0  
3           0  
4           1

In [55]:
heart_train.dtypes

gender               int64
age                  int64
education          float64
currentSmoker        int64
cigsPerDay         float64
BPMeds             float64
prevalentStroke      int64
prevalentHyp         int64
diabetes             int64
totChol            float64
sysBP              float64
diaBP              float64
BMI                float64
heartRate          float64
glucose            float64
TenYearCHD           int64
dtype: object

In [56]:
heart_test.dtypes

gender               int64
age                  int64
education          float64
currentSmoker        int64
cigsPerDay         float64
BPMeds             float64
prevalentStroke      int64
prevalentHyp         int64
diabetes             int64
totChol            float64
sysBP              float64
diaBP              float64
BMI                float64
heartRate          float64
glucose            float64
TenYearCHD           int64
dtype: object

Checking missing values on both heart_train and heart_test

In [57]:
heart_train.isna().sum()

gender               0
age                  0
education           84
currentSmoker        0
cigsPerDay          25
BPMeds              42
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             44
sysBP                0
diaBP                0
BMI                 16
heartRate            1
glucose            315
TenYearCHD           0
dtype: int64

In [58]:
heart_test.isna().sum()

gender              0
age                 0
education          21
currentSmoker       0
cigsPerDay          4
BPMeds             11
prevalentStroke     0
prevalentHyp        0
diabetes            0
totChol             6
sysBP               0
diaBP               0
BMI                 3
heartRate           0
glucose            73
TenYearCHD          0
dtype: int64

Checking the shape

In [59]:
heart_train.shape

(3390, 16)

In [60]:
heart_test.shape

(848, 16)

In [61]:
heart_train['education'].unique()

array([ 1.,  3.,  2.,  4., nan])

In [62]:
heart_test['education'].unique()

array([ 1.,  2.,  3., nan,  4.])

In [63]:
heart_train['cigsPerDay'].unique()

array([ 0.,  2., 30.,  3., 20., 15., 35., 23., 10.,  1., nan, 11., 40.,
        5.,  9., 17., 16., 43., 12., 25.,  7.,  6.,  4., 50.,  8., 60.,
       18., 70., 13., 38., 14., 45., 19.])

In [64]:
heart_test['cigsPerDay'].unique()

array([ 0., 23.,  1., 20., 30.,  7., 10.,  3.,  5.,  9., 43., 40., 15.,
        6., 35., 25., 13.,  2., 60., 18., nan, 11., 29., 12., 45.,  8.,
       17.])

In [65]:
heart_train['BPMeds'].unique()

array([ 0.,  1., nan])

In [66]:
heart_test['BPMeds'].unique()

array([ 0.,  1., nan])

Checking the statistics of both the train and test set

In [67]:
heart_train.describe()

gender          age    education  currentSmoker   cigsPerDay  \
count  3390.000000  3390.000000  3306.000000    3390.000000  3365.000000   
mean      0.433628    49.614159     1.981246       0.486726     8.962556   
std       0.495648     8.609617     1.023009       0.499897    12.030447   
min       0.000000    32.000000     1.000000       0.000000     0.000000   
25%       0.000000    42.000000     1.000000       0.000000     0.000000   
50%       0.000000    49.000000     2.000000       0.000000     0.000000   
75%       1.000000    57.000000     3.000000       1.000000    20.000000   
max       1.000000    70.000000     4.000000       1.000000    70.000000   

            BPMeds  prevalentStroke  prevalentHyp     diabetes      totChol  \
count  3348.000000      3390.000000   3390.000000  3390.000000  3346.000000   
mean      0.028375         0.004720      0.302065     0.024779   236.759713   
std       0.166067         0.068548      0.459221     0.155473    45.347259   
min       0.000000         0.000000      0.000000     0.000000   107.000000   
25%       0.000000         0.000000      0.000000     0.000000   206.000000   
50%       0.000000         0.000000      0.000000     0.000000   233.500000   
75%       0.000000         0.000000      1.000000     0.000000   264.000000   
max       1.000000         1.000000      1.000000     1.000000   696.000000   

             sysBP        diaBP          BMI    heartRate      glucose  \
count  3390.000000  3390.000000  3374.000000  3389.000000  3075.000000   
mean    132.126696    82.752065    25.793874    75.776335    82.023089   
std      22.047822    11.933836     4.045690    11.952094    23.934546   
min      83.500000    50.000000    15.540000    44.000000    40.000000   
25%     116.500000    74.500000    23.080000    68.000000    71.000000   
50%     128.000000    82.000000    25.425000    75.000000    78.000000   
75%     143.500000    89.000000    28.060000    82.000000    87.000000   
max     295.000000   142.500000    56.800000   143.000000   394.000000   

        TenYearCHD  
count  3390.000000  
mean      0.152212  
std       0.359280  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000

In [68]:
heart_test.describe()

gender         age   education  currentSmoker  cigsPerDay  \
count  848.000000  848.000000  827.000000     848.000000  844.000000   
mean     0.411557   49.468160    1.969770       0.523585    9.164692   
std      0.492406    8.424758    1.007385       0.499738   11.475080   
min      0.000000   33.000000    1.000000       0.000000    0.000000   
25%      0.000000   42.000000    1.000000       0.000000    0.000000   
50%      0.000000   49.000000    2.000000       1.000000    1.500000   
75%      1.000000   56.000000    3.000000       1.000000   20.000000   
max      1.000000   69.000000    4.000000       1.000000   60.000000   

           BPMeds  prevalentStroke  prevalentHyp    diabetes     totChol  \
count  837.000000       848.000000    848.000000  848.000000  842.000000   
mean     0.034648         0.010613      0.344340    0.029481  236.570071   
std      0.182995         0.102533      0.475433    0.169251   41.471282   
min      0.000000         0.000000      0.000000    0.000000  145.000000   
25%      0.000000         0.000000      0.000000    0.000000  206.000000   
50%      0.000000         0.000000      0.000000    0.000000  235.000000   
75%      0.000000         0.000000      1.000000    0.000000  261.750000   
max      1.000000         1.000000      1.000000    1.000000  453.000000   

            sysBP       diaBP         BMI   heartRate     glucose  TenYearCHD  
count  848.000000  848.000000  845.000000  848.000000  775.000000  848.000000  
mean   133.254717   83.458726   25.834485   76.288915   81.743226    0.150943  
std     21.989009   11.808590    4.217051   12.318492   24.074923    0.358205  
min     92.000000   48.000000   16.750000   45.000000   40.000000    0.000000  
25%    117.875000   75.500000   23.060000   67.750000   71.000000    0.000000  
50%    129.000000   82.500000   25.310000   75.000000   77.000000    0.000000  
75%    144.500000   90.000000   27.930000   85.000000   86.000000    0.000000  
max    215.000000  140.000000   51.280000  125.000000  325.000000    1.000000

### Imputing Missing Values in heart_train and heart_test

Data Imputation: Using subclass methods to handle the missing values. 
For numerical features, we could use mean or median imputation, while for categorical features, mode imputation is suitable.

In order to fill the missing values with sub class, we need to find what features are related with each other the most. For that, we have used the correlation matrix to identify features most strongly correlated with the column containing missing values.

In [69]:
heart_train.corr()

gender       age  education  currentSmoker  cigsPerDay  \
gender           1.000000 -0.021699   0.017740       0.197109    0.312110   
age             -0.021699  1.000000  -0.156948      -0.218453   -0.186060   
education        0.017740 -0.156948   1.000000       0.011334    0.003289   
currentSmoker    0.197109 -0.218453   0.011334       1.000000    0.771015   
cigsPerDay       0.312110 -0.186060   0.003289       0.771015    1.000000   
BPMeds          -0.063121  0.126578  -0.002784      -0.047996   -0.043402   
prevalentStroke  0.000538  0.068083  -0.032811      -0.032615   -0.029953   
prevalentHyp    -0.003935  0.304355  -0.062608      -0.096925   -0.063781   
diabetes         0.021348  0.110090  -0.044612      -0.048919   -0.046534   
totChol         -0.075607  0.262306  -0.009843      -0.051385   -0.021426   
sysBP           -0.041278  0.386243  -0.113493      -0.130769   -0.090941   
diaBP            0.057516  0.204637  -0.043689      -0.103222   -0.055125   
BMI              0.090680  0.142417  -0.122646      -0.164121   -0.091066   
heartRate       -0.117163 -0.018992  -0.048114       0.057198    0.078196   
glucose          0.017359  0.121737  -0.033716      -0.057929   -0.067160   
TenYearCHD       0.093203  0.215881  -0.036786       0.022754    0.054683   

                   BPMeds  prevalentStroke  prevalentHyp  diabetes   totChol  \
gender          -0.063121         0.000538     -0.003935  0.021348 -0.075607   
age              0.126578         0.068083      0.304355  0.110090  0.262306   
education       -0.002784        -0.032811     -0.062608 -0.044612 -0.009843   
currentSmoker   -0.047996        -0.032615     -0.096925 -0.048919 -0.051385   
cigsPerDay      -0.043402        -0.029953     -0.063781 -0.046534 -0.021426   
BPMeds           1.000000         0.069341      0.260374  0.066023  0.098285   
prevalentStroke  0.069341         1.000000      0.085928  0.016710  0.008202   
prevalentHyp     0.260374         0.085928      1.000000  0.072848  0.169543   
diabetes         0.066023         0.016710      0.072848  1.000000  0.053439   
totChol          0.098285         0.008202      0.169543  0.053439  1.000000   
sysBP            0.253094         0.047633      0.696449  0.115293  0.215922   
diaBP            0.197538         0.041469      0.619885  0.035596  0.166074   
BMI              0.090651         0.031158      0.292622  0.077206  0.117041   
heartRate        0.013199        -0.022485      0.147365  0.034112  0.076157   
glucose          0.062359         0.015608      0.076927  0.603499  0.053876   
TenYearCHD       0.103217         0.078652      0.168354  0.085651  0.087134   

                    sysBP     diaBP       BMI  heartRate   glucose  TenYearCHD  
gender          -0.041278  0.057516  0.090680  -0.117163  0.017359    0.093203  
age              0.386243  0.204637  0.142417  -0.018992  0.121737    0.215881  
education       -0.113493 -0.043689 -0.122646  -0.048114 -0.033716   -0.036786  
currentSmoker   -0.130769 -0.103222 -0.164121   0.057198 -0.057929    0.022754  
cigsPerDay      -0.090941 -0.055125 -0.091066   0.078196 -0.067160    0.054683  
BPMeds           0.253094  0.197538  0.090651   0.013199  0.062359    0.103217  
prevalentStroke  0.047633  0.041469  0.031158  -0.022485  0.015608    0.078652  
prevalentHyp     0.696449  0.619885  0.292622   0.147365  0.076927    0.168354  
diabetes         0.115293  0.035596  0.077206   0.034112  0.603499    0.085651  
totChol          0.215922  0.166074  0.117041   0.076157  0.053876    0.087134  
sysBP            1.000000  0.787136  0.331848   0.188739  0.138556    0.206707  
diaBP            0.787136  1.000000  0.377363   0.178134  0.050901    0.138634  
BMI              0.331848  0.377363  1.000000   0.061650  0.081233    0.060039  
heartRate        0.188739  0.178134  0.061650   1.000000  0.090087    0.028420  
glucose          0.138556  0.050901  0.081233   0.090087  1.000000    0.101690  
TenYearCHD       0.206707  0.138634  0.060039   0.02

Using the sysBP and diaBP, for that categorizing systolic and diastolic BP into ranges.\
For SysBP : Values < 120 are low, 120-140 are normal and >120 high\
For DiaBP : Values < 80 are low, 80-90 are normal and >90 high

In [70]:
def categorize_bp(df, sys_threshold=140, dia_threshold=90):
    """Categorize systolic and diastolic BP into ranges."""
    sys_category = pd.cut(df['sysBP'], bins=[-float('inf'), 120, sys_threshold, float('inf')],
                          labels=['low', 'normal', 'high'], right=True)
    dia_category = pd.cut(df['diaBP'], bins=[-float('inf'), 80, dia_threshold, float('inf')],
                          labels=['low', 'normal', 'high'], right=True)
    return sys_category, dia_category

#### For heart_test, we will ensure that imputation is based on statistics derived from heart_train in order to avoid data leakage.

##### Imputation Logic for Education, cigsPerDay, and Glucose. 

In [71]:
def MissingValueImputer_1(train_df, test_df):
    # Education: filling with mode (categorical variable)
    train_df['education'].fillna(train_df['education'].mode()[0], inplace=True)
    test_df['education'].fillna(train_df['education'].mode()[0], inplace=True)

    # Smoker: filling with the median of current smokers (positively skewed, so using median)
    # Set to 0 for non-smokers as they don't smoke
    train_df.loc[train_df['currentSmoker'] == 0, 'cigsPerDay'] = 0
    test_df.loc[test_df['currentSmoker'] == 0, 'cigsPerDay'] = 0
    
    median_cigs = train_df[train_df['currentSmoker'] == 1]['cigsPerDay'].median()
    
    train_df.loc[train_df['currentSmoker'] == 1, 'cigsPerDay'] = train_df.loc[train_df['currentSmoker'] == 1, 'cigsPerDay'].fillna(median_cigs)
    test_df.loc[test_df['currentSmoker'] == 1, 'cigsPerDay'] = test_df.loc[test_df['currentSmoker'] == 1, 'cigsPerDay'].fillna(median_cigs)

    # Glucose: based on diabetes status (max is significantly higher than 75th percentile, using median)
    # Highest corelation with diabetes columns
    train_df['glucose'] = train_df.groupby('diabetes', observed=False)['glucose'].transform(lambda x: x.fillna(x.median()))
    test_df['glucose'] = test_df.groupby('diabetes', observed=False)['glucose'].transform(
        lambda x: x.fillna(train_df.groupby('diabetes', observed=False)['glucose'].median()[x.name]))
    
    return train_df, test_df

##### Imputation Logic for BPMeds. Having strong corelation with prevalentHyp,sysBP,diaBP

In [72]:
def MissingValueImputer_2(train_df, test_df):
    # BP Meds: 
    # Rule 1: Patients with prevalent hypertension (prevalentHyp == 1)
    train_df.loc[train_df['BPMeds'].isnull() & (train_df['prevalentHyp'] == 1), 'BPMeds'] = 1
    test_df.loc[test_df['BPMeds'].isnull() & (test_df['prevalentHyp'] == 1), 'BPMeds'] = 1
    
    # Rule 2: Patients with very high systolic blood pressure (sysBP > 140) and no prevalent hypertension
    high_bp_threshold = 140
    train_df.loc[train_df['BPMeds'].isnull() & (train_df['sysBP'] > high_bp_threshold), 'BPMeds'] = 1
    test_df.loc[test_df['BPMeds'].isnull() & (test_df['sysBP'] > high_bp_threshold), 'BPMeds'] = 1

    # Rule 3: Patients with very high diaBP blood pressure (diaBP > 90) and no prevalent hypertension
    train_df.loc[train_df['BPMeds'].isnull() & (train_df['diaBP'] > 90), 'BPMeds'] = 1
    test_df.loc[test_df['BPMeds'].isnull() & (test_df['diaBP'] > 90), 'BPMeds'] = 1

    # Rule 4: Remaining missing values default to 0 (not on medication)
    train_df['BPMeds'].fillna(0, inplace=True)
    test_df['BPMeds'].fillna(0, inplace=True)
    train_df['BPMeds'] = train_df['BPMeds'].astype(int)
    test_df['BPMeds'] = test_df['BPMeds'].astype(int)
    
    return train_df, test_df

##### Imputation Logic for Cholestrol, BMI, and Heart Rate.

Using subclass(based on corelation matrix which are highly corelated) and considering median for it. After reading the statistics we got in describe method, we saw feature cholestrol is positively skewed, therefore using median as it is robust to outliers.

For Heart rate also, using median based on the statistics, as max is significantly higher than 75th percentile. Using Median as it is robust to outliers. Same for BMI as well, mean is higher than median and max is higher than 7th percentile. 

In [73]:
def MissingValueImputer_3(train_df, test_df):
    
    # Categorize sysBP for train and test datasets
    train_df['sysBP_category'], train_df['diaBP_category'] = categorize_bp(train_df)
    test_df['sysBP_category'], test_df['diaBP_category'] = categorize_bp(test_df)

    # Cholesterol: Filling based on age and sysBP_category (Highest corelation with these columns)
    train_df['totChol'] = train_df.groupby(['age', 'sysBP_category'], observed=False)['totChol'].transform(
        lambda x: x.fillna(x.median())
    )
    test_df['totChol'] = test_df.groupby(['age', 'sysBP_category'], observed=False)['totChol'].transform(
        lambda x: x.fillna(train_df.groupby(['age', 'sysBP_category'], observed=False)['totChol'].median()[x.name])
    )
    
    # BMI: filling by using age, gender, prevalentHyp, sysBP, and diaBP (Highest corelation with these columns)
    train_df['BMI'] = train_df.groupby(['age', 'gender', 'prevalentHyp', 'sysBP_category', 'diaBP_category'], observed=False)['BMI'].transform(
        lambda x: x.fillna(x.median() if not x.isnull().all() else train_df['BMI'].median()))
    test_df['BMI'] = test_df.groupby(['age', 'gender', 'prevalentHyp', 'sysBP_category', 'diaBP_category'], observed=False)['BMI'].transform(
        lambda x: x.fillna(train_df.groupby(['age', 'gender', 'prevalentHyp', 'sysBP_category', 'diaBP_category'], observed=False)['BMI'].median().get(x.name, train_df['BMI'].median()) if not x.isnull().all() else train_df['BMI'].median()))

    # Heart Rate: filling by using sysBP and diaBP (Highest corelation with these columns)
    train_df['heartRate'] = train_df.groupby(['sysBP_category', 'diaBP_category'], observed=False)['heartRate'].transform(lambda x: x.fillna(x.median()))
    test_df['heartRate'] = test_df.groupby(['sysBP_category', 'diaBP_category'], observed=False)['heartRate'].transform(
        lambda x: x.fillna(train_df.groupby(['sysBP_category', 'diaBP_category'], observed=False)['heartRate'].median()[x.name]))

    # Dropping the temporary sysBP_category column
    train_df.drop(columns=['sysBP_category', 'diaBP_category'], inplace=True)
    test_df.drop(columns=['sysBP_category','diaBP_category'], inplace=True)

    return train_df, test_df

In [74]:
# Imputation Logic for Education, cigsPerDay, and Glucose
MissingValueImputer_1(heart_train, heart_test)
# Imputation Logic for BPMeds¶
MissingValueImputer_2(heart_train, heart_test)
# Imputation Logic for Cholestrol, BMI, and Heart Rate
MissingValueImputer_3(heart_train, heart_test)

print("Train Data Missing Values After Imputation:\n", heart_train.isnull().sum())
print("\nTest Data Missing Values After Imputation:\n", heart_test.isnull().sum())

/var/folders/f9/tzl1jwjn3f3fvhyb_dm77j6h0000gn/T/ipykernel_85751/3040251283.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  train_df['education'].fillna(train_df['education'].mode()[0], inplace=True)
/var/folders/f9/tzl1jwjn3f3fvhyb_dm77j6h0000gn/T/ipykernel_85751/3040251283.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which

Train Data Missing Values After Imputation:
 gender             0
age                0
education          0
currentSmoker      0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
TenYearCHD         0
dtype: int64

Test Data Missing Values After Imputation:
 gender             0
age                0
education          0
currentSmoker      0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
TenYearCHD         0
dtype: int64


### Normalising the data

Applying z-score normalization on heart_train, and then normalizing heart_test using the mean and standard deviation from heart_train.

In [75]:
mean_std = {}

In [76]:
def z_score_normalize(train_df):
    for column in train_df.columns:
        if train_df[column].dtype in ['int64', 'float64'] and column != 'TenYearCHD':
            mean = train_df[column].mean()
            std = train_df[column].std()
            mean_std[column] = (mean, std)
            train_df[column] = (train_df[column] - mean) / std

In [77]:
z_score_normalize(heart_train)

In [78]:
print(mean_std)

{'gender': (0.4336283185840708, 0.495648331104533), 'age': (49.614159292035396, 8.609616730841537), 'education': (1.9569321533923303, 1.0217048588742135), 'currentSmoker': (0.48672566371681414, 0.4998974975685707), 'cigsPerDay': (9.043952802359883, 12.023146397122822), 'BPMeds': (0.033628318584070796, 0.18029709890002746), 'prevalentStroke': (0.00471976401179941, 0.06854833286068387), 'prevalentHyp': (0.30206489675516224, 0.4592209735457691), 'diabetes': (0.024778761061946902, 0.15547316301364433), 'totChol': (236.71312684365782, 45.08705642840767), 'sysBP': (132.12669616519173, 22.047822389406125), 'diaBP': (82.75206489675516, 11.933836490922324), 'BMI': (25.78600442477876, 4.039053065029307), 'heartRate': (75.77610619469027, 11.95033818141861), 'glucose': (81.70766961651917, 22.89993435334305)}


In [79]:
def apply_normalization(heart_test):
    for column, (mean, std) in mean_std.items():
        heart_test[column] = (heart_test[column] - mean) / std

In [80]:
apply_normalization(heart_test)

In [81]:
heart_train.head()

gender       age  education  currentSmoker  cigsPerDay    BPMeds  \
0  1.142689  1.554755  -0.936603      -0.973651   -0.752212 -0.186516   
1  1.142689 -0.535931   1.020909      -0.973651   -0.752212 -0.186516   
2 -0.874871  0.160964  -0.936603       1.026759   -0.585866 -0.186516   
3  1.142689 -0.535931   1.020909       1.026759    1.742975 -0.186516   
4 -0.874871 -0.535931   0.042153       1.026759   -0.502693 -0.186516   

   prevalentStroke  prevalentHyp  diabetes   totChol     sysBP     diaBP  \
0        -0.068853      1.519824 -0.159376 -1.036065  0.719949  0.607343   
1        -0.068853     -0.657777 -0.159376 -1.657086 -0.323238  0.523548   
2        -0.068853     -0.657777 -0.159376  0.538666 -0.232526 -0.146815   
3        -0.068853     -0.657777 -0.159376  0.294694 -0.277882  0.565446   
4        -0.068853     -0.657777 -0.159376  0.294694 -0.096458 -0.230610   

        BMI  heartRate   glucose  TenYearCHD  
0  0.332750  -0.315983  0.187439           0  
1  0.543196   0.771852  0.056434           0  
2 -1.373095  -0.064944  0.624121           0  
3  0.716503  -0.064944  0.449448           0  
4 -1.373095   1.190250  0.187439           0

In [82]:
heart_test.head()

gender       age  education  currentSmoker  cigsPerDay    BPMeds  \
0 -0.874871  0.741710  -0.936603      -0.973651   -0.752212 -0.186516   
1 -0.874871 -0.419782   0.042153      -0.973651   -0.752212 -0.186516   
2 -0.874871  0.277113  -0.936603      -0.973651   -0.752212 -0.186516   
3 -0.874871 -0.419782   1.020909       1.026759    1.160765 -0.186516   
4  1.142689  2.251650  -0.936603       1.026759   -0.669039 -0.186516   

   prevalentStroke  prevalentHyp  diabetes   totChol     sysBP     diaBP  \
0        -0.068853      1.519824  6.272602  0.804818  0.175677 -0.230610   
1        -0.068853     -0.657777 -0.159376  0.294694 -0.504662 -0.146815   
2        -0.068853     -0.657777 -0.159376  0.849177 -0.912865 -0.984768   
3        -0.068853     -0.657777 -0.159376  1.070970 -0.096458  0.104571   
4        -0.068853     -0.657777 -0.159376  0.183797 -0.413950 -0.481996   

        BMI  heartRate   glucose  TenYearCHD  
0  0.481300   1.190250  5.602301           1  
1  0.728883   1.608649 -0.249244           0  
2 -0.026245   0.353454 -0.161907           0  
3 -0.665008   0.771852  0.143770           0  
4  0.196580  -0.483342 -0.030903           1

Our both datasets are free from missing values and are in normalised form now. We can begin to build logistic regression model.

In [83]:
# Preparing data for training and testing by separate features and target variable
X_train = heart_train.drop(columns=['TenYearCHD']).values
y_train = heart_train['TenYearCHD'].values

X_test = heart_test.drop(columns=['TenYearCHD']).values
y_test = heart_test['TenYearCHD'].values

In [84]:
# Adding bias term to X
X_train = np.c_[np.ones(X_train.shape[0]), X_train]
X_test = np.c_[np.ones(X_test.shape[0]), X_test]

In [85]:
# Initialize weights and setting other parameters
# initializing with the same logic as in ques 1
weights = np.random.randn(X_train.shape[1]) * 0.01

m, n = X_train.shape
learning_rate = 0.1
epochs = 10

<img src= "attachment:3c7b316e-2a0c-4fe5-ac27-93c26c6c0813.png" width="400">

Sigmoid function

In [86]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

Cost function (Binary Cross-Entropy)

In [87]:
def compute_cost(y, y_pred):
    m = len(y)
    cost = -(1 / m) * np.sum(y * np.log(y_pred) + (1 - y) * np.log(1 - y_pred))
    return cost

Gradient Descent Algorithm based on Logistic Regression

From class slides:

<img src= "attachment:099f45c6-1cc4-41e3-b5ad-39da120dd97e.png" width="300">

In [88]:
def gradient_descent(X, y, weights, learning_rate, epochs):
    m = X.shape[0]
    for epoch in range(epochs):
        # z as per formula
        z = np.dot(X, weights)
        y_pred = sigmoid(z)
        
        # Compute cost
        cost = compute_cost(y, y_pred)
        print(f"Epoch {epoch+1}, Cost: {cost}")
        
        # Compute gradients
        dw = (1 / m) * np.dot(X.T, (y_pred - y))
        
        # Update weights
        weights -= learning_rate * dw
    
    return weights

Prediction

In [89]:
def predict(X, weights):
    z = np.dot(X, weights)
    y_pred = sigmoid(z)
    return [1 if i >= 0.5 else 0 for i in y_pred]

Please print out the Cost function J for every iteration (until 10 epochs) for the train dataset.

Train the model

In [90]:
optimal_weights = gradient_descent(X_train, y_train, weights, learning_rate, epochs)

Epoch 1, Cost: 0.6939275939216488
Epoch 2, Cost: 0.6792177301857384
Epoch 3, Cost: 0.66546804063382
Epoch 4, Cost: 0.6525948555863147
Epoch 5, Cost: 0.640523025413177
Epoch 6, Cost: 0.6291851594335296
Epoch 7, Cost: 0.618520860227914
Epoch 8, Cost: 0.6084759855469655
Epoch 9, Cost: 0.5990019574596301
Epoch 10, Cost: 0.5900551291807082


Make predictions on train and test sets

In [91]:
train_predictions = predict(X_train, optimal_weights)
test_predictions = predict(X_test, optimal_weights)

##### Evaluation

Please print out confusion matrix and its corresponding accuracy, precision, and recall for the train and test dataset after the last iteration. 

##### Confusion Matrix

In [92]:
def evaluate(y_true, y_pred):
    tp = sum((y_true == 1) & (np.array(y_pred) == 1))
    tn = sum((y_true == 0) & (np.array(y_pred) == 0))
    fp = sum((y_true == 0) & (np.array(y_pred) == 1))
    fn = sum((y_true == 1) & (np.array(y_pred) == 0))

    print("\nConfusion Matrix:")
    print("----------------")
    print(f"            Predicted")
    print(f"             1    0")
    print(f"Actual  1  [TP:{tp:3d} FN:{fn:3d}]")
    print(f"        0  [FP:{fp:3d} TN:{tn:3d}]")
    
    accuracy = (tp + tn) / len(y_true)
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0
    
    print("\nMetrics:")
    print(f"Accuracy:  {accuracy:.3f}")
    print(f"Precision: {precision:.3f}")
    print(f"Recall:    {recall:.3f}")

In [93]:
# Evaluate on train dataset
print("\nTrain Dataset Evaluation:")
evaluate(y_train, train_predictions)


Train Dataset Evaluation:

Confusion Matrix:
----------------
            Predicted
             1    0
Actual  1  [TP: 79 FN:437]
        0  [FP:100 TN:2774]

Metrics:
Accuracy:  0.842
Precision: 0.441
Recall:    0.153


In [94]:
# Evaluate on test dataset
print("\nTest Dataset Evaluation:")
evaluate(y_test, test_predictions)


Test Dataset Evaluation:

Confusion Matrix:
----------------
            Predicted
             1    0
Actual  1  [TP: 23 FN:105]
        0  [FP: 25 TN:695]

Metrics:
Accuracy:  0.847
Precision: 0.479
Recall:    0.180


The model shows an accuracy of approximately 83-84% on both the train and test datasets. However, this seems to be misleading, if we notice the precision and recall metrics for both train and test datasets. On the train set, the precision is 0.441 and recall is 0.153, while on the test set, precision improves slightly to 0.479 and recall increases to 0.180. This tells that the model is more effective at reducing false positives (higher precision) but struggles with capturing a good portion of the actual positive cases (relatively low recall).

This low recall tells also tells us that dataset is imbalanced. In such cases, model predicts well on majority class(here '0') and struggles to classify minority class(here '1'). The model is maintaining high overall accuracy due to correct prediction of the majority class. The low recall tells us that model is missing many of the positive cases. 

Original Requirement for the assignment is finished at this point. Trying additional things to improve the results.

### =============================================

## Trying to Address imbalance by using weighted cost function

Applying higher weights to the minority class and lower to majority class to allow balance for the dataset. Updated the cost function to use the calculated weights.

In [95]:
def compute_cost_w(y, y_pred, w1, w0):
    m = len(y)
    cost = - (1 / m) * np.sum(
        w1 * y * np.log(y_pred + 1e-15) + 
        w0 * (1 - y) * np.log(1 - y_pred + 1e-15)
    )
    return cost

In [96]:
def gradient_descent_w(X, y, weights, learning_rate, epochs):
    m = len(y)
    
    pos_count = np.sum(y)
    neg_count = m - pos_count
    w1 = m / (2 * pos_count)
    w0 = m / (2 * neg_count)
    
    for epoch in range(epochs):
        z = np.dot(X, weights)
        y_pred = sigmoid(z)
        
        cost = compute_cost_w(y, y_pred, w1, w0)
        print(f"Epoch {epoch+1}, Cost: {cost}")
        dw = (1 / m) * np.dot(X.T, (y_pred - y))
        
        weights -= learning_rate * dw
    
    return weights

In [97]:
def predict_w(X, weights):
    z = np.dot(X, weights)
    y_pred = sigmoid(z)
    return [1 if i >= 0.5 else 0 for i in y_pred]

In [98]:
weights = np.random.randn(X_train.shape[1]) * 0.01

m, n = X_train.shape
learning_rate = 0.1
epochs = 10

optimal_weights_w = gradient_descent_w(X_train, y_train, weights, learning_rate, epochs)

Epoch 1, Cost: 0.6924977446891223
Epoch 2, Cost: 0.6878514814752195
Epoch 3, Cost: 0.6838852667699582
Epoch 4, Cost: 0.6805289367686805
Epoch 5, Cost: 0.6777198992467847
Epoch 6, Cost: 0.6754023698599398
Epoch 7, Cost: 0.6735266395452729
Epoch 8, Cost: 0.6720483916123675
Epoch 9, Cost: 0.670928078135847
Epoch 10, Cost: 0.6701303589975697


In [99]:
train_predictions_w = predict_w(X_train, optimal_weights_w)
test_predictions_w = predict_w(X_test, optimal_weights_w)

In [100]:
print("\nTrain Dataset Evaluation:")
evaluate(y_train, train_predictions_w)


Train Dataset Evaluation:

Confusion Matrix:
----------------
            Predicted
             1    0
Actual  1  [TP:107 FN:409]
        0  [FP:143 TN:2731]

Metrics:
Accuracy:  0.837
Precision: 0.428
Recall:    0.207


In [101]:
print("\nTest Dataset Evaluation:")
evaluate(y_test, test_predictions_w)


Test Dataset Evaluation:

Confusion Matrix:
----------------
            Predicted
             1    0
Actual  1  [TP: 27 FN:101]
        0  [FP: 38 TN:682]

Metrics:
Accuracy:  0.836
Precision: 0.415
Recall:    0.211


Looking at the results, we don't see any improvement over the original model which did not consider weights. Trying an alternate approach. 

## Trying to address imbalance using SMOTE method

Creating synthetic data for the minority class to balance the training dataset.

In [102]:
from imblearn.over_sampling import SMOTE
import numpy as np

smote = SMOTE(random_state=42)
X_balanced, y_balanced = smote.fit_resample(X_train, y_train)

In [103]:
weights = np.zeros(X_balanced.shape[1])
learning_rate = 0.01
epochs = 10

trained_weights = gradient_descent(X_balanced, y_balanced, weights, learning_rate, epochs)

Epoch 1, Cost: 0.6931471805599453
Epoch 2, Cost: 0.6921968127898235
Epoch 3, Cost: 0.6912618768164219
Epoch 4, Cost: 0.6903420909159323
Epoch 5, Cost: 0.6894371777202472
Epoch 6, Cost: 0.6885468642193586
Epoch 7, Cost: 0.6876708817590796
Epoch 8, Cost: 0.6868089660343142
Epoch 9, Cost: 0.6859608570781046
Epoch 10, Cost: 0.6851262992466836


In [104]:
def predict_sm(X, weights, threshold=0.5):
    z = np.dot(X, weights)
    y_pred = sigmoid(z)
    return (y_pred >= threshold).astype(int)

In [105]:
train_predictions_sm = predict_sm(X_balanced, trained_weights)
test_predictions_sm = predict_sm(X_test, trained_weights)

In [106]:
print("\nTrain Dataset Evaluation:")
evaluate(y_balanced, train_predictions_sm)


Train Dataset Evaluation:

Confusion Matrix:
----------------
            Predicted
             1    0
Actual  1  [TP:1945 FN:929]
        0  [FP:1103 TN:1771]

Metrics:
Accuracy:  0.646
Precision: 0.638
Recall:    0.677


In [107]:
print("\nTest Dataset Evaluation:")
evaluate(y_test, test_predictions_sm)


Test Dataset Evaluation:

Confusion Matrix:
----------------
            Predicted
             1    0
Actual  1  [TP: 96 FN: 32]
        0  [FP:291 TN:429]

Metrics:
Accuracy:  0.619
Precision: 0.248
Recall:    0.750


Results seem promising on the Train dataset with high precision and recall. However, low precision on test dataset means the model is overfitting. Further exploration use regularization techniques can be evaluated. 

## Thank you!